adapters/repos/db/vector/hnsw/search.go

```diff
@@ -226,28 +232,109 @@ func (h *hnsw) searchLayerByVectorWithDistancer(queryVector []float32,
-			// can no longer retrieve it from the pool. Instead we need to fallback
-			// to allocating a new slice.
-			//
-			// This was discovered as part of
-			// https://github.com/weaviate/weaviate/issues/1897
+		var slice *common.VectorUint64Slice
+		var slice2 *common.VectorUint64Slice
+		var slice3 *common.VectorUint64Slice
+		if allowList == nil || !h.acornSearch.Load() {
 			connectionsReusable = make([]uint64, len(candidateNode.connections[level]))
+			copy(connectionsReusable, candidateNode.connections[level])
 		} else {
-			connectionsReusable = connectionsReusable[:len(candidateNode.connections[level])]
-		}
+			slice = h.pools.tempVectorsUint64.Get(M * h.maximumConnectionsLayerZero)
+			connectionsReusable = slice.Slice
+			slice2 = h.pools.tempVectorsUint64.Get(h.maximumConnectionsLayerZero)
+			slice3 = h.pools.tempVectorsUint64.Get(M * h.maximumConnectionsLayerZero * h.maximumConnectionsLayerZero * h.maximumConnectionsLayerZero)
+			pendingNextRound := slice2.Slice
+			pendingThisRound := slice3.Slice
+
+			realLen := 0
+			index := 0
+
+			pendingNextRound = pendingNextRound[:len(candidateNode.connections[level])]
+			copy(pendingNextRound, candidateNode.connections[level])
+			for realLen < M*h.maximumConnectionsLayerZero && len(pendingNextRound) > 0 {
+				if cap(pendingThisRound) >= len(pendingNextRound) {
+					pendingThisRound = pendingThisRound[:len(pendingNextRound)]
+				} else {
+					pendingThisRound = make([]uint64, len(pendingNextRound))
+					slice3.Slice = pendingThisRound
+				}
+				copy(pendingThisRound, pendingNextRound)
+				pendingNextRound = pendingNextRound[:0]
+				for index < len(pendingThisRound) && realLen < M*h.maximumConnectionsLayerZero {
+					nodeId := pendingThisRound[index]
+					index++
+					if ok := visited.Visited(nodeId); ok {
+						// skip if we've already visited this neighbor
+						continue
+					}
+					if !visitedExp.Visited(nodeId) {
+						if allowList.Contains(nodeId) {
+							connectionsReusable[realLen] = nodeId
+							realLen++
+							visitedExp.Visit(nodeId)
+							continue
+						}
+					} else {
+						continue
+					}
+					visitedExp.Visit(nodeId)
 
-		copy(connectionsReusable, candidateNode.connections[level])
+					node := h.nodes[nodeId]
+					if node == nil {
+						continue
+					}
+					for _, expId := range node.connections[level] {
+						if visitedExp.Visited(expId) {
+							continue
+						}
+						if visited.Visited(expId) {
+							continue
+						}
+
+						if realLen >= M*h.maximumConnectionsLayerZero {
+							break
+						}
+
+						if allowList.Contains(expId) {
+							visitedExp.Visit(expId)
+							connectionsReusable[realLen] = expId
+							realLen++
+						} else {
+							visitedExp.Visit(expId)
+							pendingNextRound = append(pendingNextRound, expId)
+						}
+					}
+				}
+			}
+			if realLen == 0 && results.Len() < 10 {
+				now := []uint64{pendingThisRound[0]}
+				var next []uint64
+				pass := false
+				hops := 0
+				for !pass {
```

Comment:
How does this exit if `allowList.Contains(y)` is never true?

Reply:
I have added a second check. We never do more than three hops.